### PR TITLE
Port Cortex to use Prometheus 2 packages

### PR DESCRIPTION
**Pull request by juliusv**
This is the main part of https://github.com/weaveworks/cortex/issues/555

As promised, this is a crazy mega change. Unfortunately, when swapping out vendored packages and making everything work again, a smaller change wasn't really possible.

Here's a high-level overview:

- We change the vendoring of most Prometheus packages (API, rules, promql, storage *interfaces*, etc.) to Prometheus 2.0, which is very different in many regards. The using packages in Cortex are adjusted to work with the new interfaces of Prom 2.
- This change doesn't actually change the internal storage of Cortex, and it continues to use Prom 1's chunk format. Because it is impossible to vendor packages both from Prom 1 and Prom 2, I hard-forked (copied) the necessary storage (chunk+metric) packages from Prometheus into `pkg/prom1/...` and rewrote all the imports for them.
- Alertmanager vendoring is also updated to the latest state, to be able to work properly with Prom 2's common dependencies (it's not possible to vendor multiple versions of the same repo).
- The latest `prometheus/common/log` does not automatically configure log level flags anymore, so there's extra code to do that in Cortex now.
- The Go builder image is updated from 1.8.3 to 1.9.1 now because I think I encountered some problems with the old one, but now I'm not sure anymore.
- The linter needs new exceptions with Go 1.9.1's `vet`.
- Prom 2's querier `Select()` method is used for both metadata and sample querying, but a remote implementation would want to know beforehand whether it needs to get full sample data or only metadata. That's why I allow parametrizing the `Queryable` (which produces a `Querier`) to tell it whether it should produce full sample or only metadata queriers, and then hand in the right kind of `Queryable` to the using packages.
- The `prometheus/common/route` package doesn't have a context function argument in its constructor anymore, but we also didn't need that anymore, so that dummy function argument is just removed.
- The multi-tentant Alertmanager needed only small adjustments to work with the updated upstream AM packages.
- I updated `golang.org/x/net/context` to native `context` in many places (was required for some things), but not everywhere yet.
- The `LazySeriesIterators` were never actually used lazily, and would have made Prom 2 usage much harder, so I removed them in favor of eagerly evaluating even queries that fuzzy-match the metric name.
- Underlying `Get()` and `Query()` functions now return `model.Matrix` instead of iterators, since the above-mentioned eager eval obviated the need for iterators again and this made things simpler.
- I converted many label and metric types to Prom 2 types (mostly, where necessary), but not all of them yet (didn't want to make the change even bigger than necessary). I changed helper functions accordingly.
- I copied and adapted the private `concreteSeries` / `concreteSeriesSet` / `errorSeriesSet` implementations from Prometheus's remote storage code to be able to return series from `Select()` that are backed by already present sample data (or errors).
- The adapation of the new appender interface in `pkg/ruler/compat.go` fixes https://github.com/weaveworks/cortex/issues/569 as a side-effect by batching up all samples from a given rule evaluation and only sending them upon `Commit()` (called at the end of a rule evaluation).
- Extreme `dep` dependency vendoring wrangling was necessary to make all the different package versions and conflicting dependencies work. Hopefully, everything fits well enough together now (at least it compiles!).

User-visible changes due to this PR:
- All PromQL 2.0 functionality (including staleness handling) should be available.
- The latest Prom 2.0 `/api/v1` functionality should be available (there weren't many (any?)) changes. Note that `/api/v2` isn't available yet, but only contains some extra endpoints which we don't want to implement anyway so far.
- Any new Alertmanager features should be available now via the AM config (and embedded UI updates).
- **NOTE**: The rule format for alerting / recording rules is still Prometheus 1.0 format! This is because Prometheus 2.0 packages still have functionality for parsing the old format, and we are still using this. (This functionality is still in Prom 2.0 to enable `promtool` to do conversions). This is great, as it allows us to keep at least one major user-visible change out of this PR, and it is easy to change the code to expect the Prom 2.0 rule format in a subsequent change.

How to review this change? Let's be honest: you probably can't :) Instead, help test it! All the pieces should be roughly there, but I expect bugs. I tested basic ingestion/querying/alerting rules/AM UI, but no edge cases or other very thorough manual testing yet.

Please help test everything you can think of in this current state and let me know what is still broken! It would be good to get this merged after we are reasonably confident that it is working, and do other code cleanups and features in later changes, since probably a lot of other stuff is blocked on this one large code change.

### CODE REVIEW

**Comment by juliusv** [comment]
(will take a look at the remaining test failures)

**Comment by juliusv** [comment]
Hmm, tests were "just" flaky, not sure if related to my changes or not.

**Comment by rade** [comment]
did you forget to `dep prune`?

**Comment by juliusv** [comment]
@rade Indeed, thanks. Still new to `dep`. Fixed. I squashed the pruning into the existing vendoring commit to not pollute the git history with the billions of extra files forever.

**Comment by rade** [comment]
> Fixed

Actually, according to [the docs](https://github.com/weaveworks/cortex#vendoring) one is meant to run `make update-vendor`. This does the `dep ensure && dep prune` but also, crucially, retains the BUILD.bazel files.

Apologies for the mis-guidance; I haven't built cortex for a while.

**Comment by rade** [comment]
> one is meant to run `make update-vendor` ... retains the BUILD.bazel files.

...though it looks like that would also retain the BUILD.bazel files of _removed_ dependencies, which doesn't seem right.

File: cmd/querier/main.go [hunk 1]
```diff
@@ -75,18 +75,24 @@ func main() {
...
-	}).WithPrefix("/api/prom/api/v1")
+	sampleQueryable := querier.NewQueryable(dist, chunkStore, false)
+	metadataQueryable := querier.NewQueryable(dist, chunkStore, true)
+
+	engine := promql.NewEngine(sampleQueryable, nil)
+	api := v1.NewAPI(
+		engine,
+		metadataQueryable,
+		querier.DummyTargetRetriever{},
+		querier.DummyAlertmanagerRetriever{},
```
Comment (jml) [inline cmd/querier/main.go]: What does it mean to give the API dummy retrievers?

File: pkg/chunk/chunk_store.go [hunk 2]
```diff
@@ -262,40 +267,52 @@ outer:
...
-			if !matcher.Match(metric[matcher.Name]) {
+			if !matcher.Matches(string(metric[model.LabelName(matcher.Name)])) {
 				continue outer
 			}
 		}
 
-		orgID, err := user.ExtractOrgID(ctx)
-		if err != nil {
-			return nil, err
+		// TODO(prom2): Does this contraption over-match?
```
Comment (jml) [inline pkg/chunk/chunk_store.go]: How would we test this?

File: pkg/chunk/chunk_store_test.go [hunk 3]
```diff
@@ -35,32 +33,33 @@ func newTestChunkStore(t *testing.T, cfg StoreConfig) *Store {
...
 }
-func (s ByFingerprint) Less(i, j int) bool {
-	return s[i].Metric().Metric.Fingerprint() < s[j].Metric().Metric.Fingerprint()
+func (bfp ByFingerprint) Less(i, j int) bool {
+	return bfp[i].Metric.Fingerprint() < bfp[j].Metric.Fingerprint()
 }
 
-// TestChunkStore_Get tests iterators are returned correctly depending on the type of query
-func TestChunkStore_Get_concrete(t *testing.T) {
+// TODO(prom2): reintroduce tests that were part of TestChunkStore_Get_lazy
```
Comment (jml) [inline pkg/chunk/chunk_store_test.go]: Given that our next step is testing, it would be good to actually reintroduce these.

File: pkg/configs/client/configs.go [hunk 4]
```diff
@@ -8,6 +8,7 @@ import (
 	"net/url"
 	"time"
 
+	gklog "github.com/go-kit/kit/log"
```
Comment (jml) [inline pkg/configs/client/configs.go]: Why use go-kit's logger?

File: pkg/querier/series_set.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+func (errSeriesSet) At() storage.Series {
+	return nil
+}
+
+func (e errSeriesSet) Err() error {
+	return e.err
+}
+
+type metadataSeriesSet struct {
```
Comment (jml) [inline pkg/querier/series_set.go]: I don't understand what this is for.

File: cmd/querier/main.go (hunk 1 above)
Comment (juliusv) [reply cmd/querier/main.go]: The normal Prometheus API can provide a list of targets (e.g. http://demo.robustperception.io:9090/api/v1/targets) and configured Alertmanagers (http://demo.robustperception.io:9090/api/v1/alertmanagers). We don't need/have this information for Cortex, so we pass in dummy implementations that just give empty lists.

File: pkg/chunk/chunk_store.go (hunk 2 above)
Comment (juliusv) [reply pkg/chunk/chunk_store.go]: This is something I still want to look at closer before merging. Mainly I extracted this code from the lazy iterators and moved it to here (for eager loading). I just need to reason through it all one more time before giving a better answer :)

File: pkg/configs/client/configs.go (hunk 4 above)
Comment (juliusv) [reply pkg/configs/client/configs.go]: Prometheus 2.0 packages use go-kit's logger interfaces everywhere now, so we need to interface with that. See this in the same file below:

```golang
// TODO(prom2): Wrap our logger into a gokit logger.
rule = rules.NewAlertingRule(r.Name, r.Expr, r.Duration, r.Labels, r.Annotations, gklog.NewNopLogger())
```

File: pkg/querier/series_set.go (hunk 5 above)
Comment (juliusv) [reply pkg/querier/series_set.go]: Whoops, I forgot to remove that type. Removed.

File: pkg/chunk/chunk_store_test.go (hunk 3 above)
Comment (juliusv) [reply pkg/chunk/chunk_store_test.go]: Ack.

**Comment by juliusv** [comment]
@rade Hm yeah,

```Makefile
update-vendor:
        dep ensure && dep prune
        git status | grep BUILD.bazel | cut -d' ' -f 5 | xargs git checkout HEAD
```

I guess preserving those BUILD.bazel files would only work now if they hadn't already been removed in previous commits. And that also means we have to manually write Bazel files for new dependencies? Or is there an automation for that? /cc @tomwilkie

**Comment by juliusv** [comment]
Ok, so I re-added the Bazel build files by running `make gazelle` and then copying some of the problematic/missing build files from `master` (see comment in Makefile about the `gazelle` target):

```
make gazelle
# Get 
git checkout master -- vendor/github.com/openzipkin/zipkin-go-opentracing/_thrift/gen-go/zipkincore/BUILD.bazel
git checkout master -- vendor/github.com/openzipkin/zipkin-go-opentracing/_thrift/gen-go/scribe/BUILD.bazel
git checkout master -- vendor/golang.org/x/crypto/curve25519/BUILD.bazel
```

It builds with Bazel now :)

File: pkg/chunk/chunk_store.go (hunk 2 above)
Comment (juliusv) [reply pkg/chunk/chunk_store.go]: So to elaborate:

The question I have around this code is not related to this PR really, as I just moved it here to be eager-loading instead of lazy-loading. The same question applies to the current `master`.

Somehow it seems to actually select only the correct series, although if I run it in my head there should be a bug: the code gets all series where the metric name matcher applies, then filters them down to only the series where the other matchers also apply. From those matching series, it builds a set of new matchers that are then used to look up the final series chunks. But those matchers should then match too many series in some cases.

So e.g. I thought this would over-match in the following scenario:

- TSDB has both `foo{a="b"}` and `foo{a="b", c="d"}`
- Query is for `{__name__=~"foo", c!="d"}`
  - (correctly this should return `foo{a="b"}`, but not `foo{a="b", c="d"})`

What I would expect to happen in the code is then:

1. it finds and matches both `foo` series in the metric name index
2. it filters out the `foo{a="b", c="d"}` series because it doesn't match the `c!="d"` matcher
3. it builds new matchers out of the remaining series's labels (`foo{a="b"}`)
4. it uses those new matchers to look up all series that match them, which should *also* match `foo{a="b", c="d"}`, since it is a superset of labels

But somehow, it does not return this wrong series though and the query result is correct. But why? Do you see what I'm missing?

Comment (juliusv) [reply pkg/chunk/chunk_store.go]: So of course I couldn't just sleep with this mystery unsolved. Many debug Printf's later, I found out the following:

- The returned series were only correct because they came from the ingesters only, and not from the chunk store.
- The chunk store actually didn't select *any* data for the above sample scenario because of a bug I introduced (fixed in latest push) that passed in `__name__` instead of the metric name into `getMetricNameMatrix()`.
- The current `master` is actually worse: rather than selecting too many series, the lazy iterator just crashes with a `nil pointer dereference` in the above scenario, as soon as there are relevant chunks in S3. This is currently happening in production as well.

So: `master` is currently broken in a crashy way (queries fail completely), whereas this PR branch is broken in that it selects too many series in an edge case. That problem would also occur in `master`, if it didn't crash before that. Thus, I am almost tending to not try and fix this unrelated bug in this PR. @jml What do you think?

**Comment by juliusv** [comment]
@jml I finished all remaining `TODO(prom2)` items now, so I'm removing the `[WIP]`. Although it would still be good if you could also do some more manual testing of everything.

Most importantly, the over-selection edge case in the chunk store is now fixed, and there are tests to cover that and other cases. The new test cases also uncovered another bug that is fixed now. I also simplified the test scenario code a bit.

I think this is ready for final testing and review.

**Comment by jml** [comment]
@juliusv Thanks! I'm not going to get a chance to do any testing today (and am away tomorrow & all week next week). Happily, @leth has volunteered to run this on dev.

**Comment by jml** [comment]
New commits LGTM. Thanks especially for the tests!

**Comment by juliusv** [comment]
Besides testing the alert notifications end-to-end one more time in dev, does anyone see any other things that we should ensure before merging? I at least have not found more issues so far, and beginning of the week would generally be good timing for deploying to prod and then dealing with any potential fallout.

**Comment by jml** [comment]
Just chatted w/ Marcus about this in person.

- end-to-end alert notifications still need to be tested
- need to clarify migration plan for rules/alert configs

We want to have an alert correctness job, but we shouldn't block the prom2
merge/prod-deploy on that.

**Comment by jml** [comment]
@leth – tagging re above message.

**Comment by juliusv** [comment]
Note that this PR doesn't change the rule format yet, so there's no
migration needed for this yet.

I'll do the (still manual) alert testing tonight.

**Comment by bboreham** [comment]
> * Remove SchemaConfig dependency from ingester
> The ingester only needs the MaxChunkAge from the SchemaConfig, and the ingester.Config is actually a better place to store and configure that information authoritatively.

Unfortunately this caused table-manager to see a zero value for `MaxChunkAge`, which causes #621
